controllers/appgroup_controller.go

```diff
@@ -70,460 +57,120 @@ type ApplicationGroupReconciler struct {
-	err = r.List(ctx, &helmReleases, listOption)
-	if err != nil {
-		r.Log.Error(err, "failed to find generated HelmRelease instance")
-		return nil, nil, err
-	}
-
-	for _, hr := range helmReleases.Items {
-		parent := hr.Name
-		if v, ok := hr.GetAnnotations()["orkestra/parent-chart"]; ok {
-			// Use the parent charts name
-			parent = v
-		}
-
-		// Add the associated conditions for that helm chart to the helm chart condition
-		// If the helm chart is a subchart, then add that to the subchart condition
-		if parent == hr.Name {
-			chartConditionMap[parent] = append(chartConditionMap[parent], hr.Status.Conditions...)
-		} else {
-			if _, ok := subChartConditionMap[parent]; !ok {
-				subChartConditionMap[parent] = make(map[string][]metav1.Condition)
-			}
-			subChartConditionMap[parent][hr.Spec.ReleaseName] = append(subChartConditionMap[parent][hr.Spec.ReleaseName], hr.Status.Conditions...)
-		}
-	}
-	return chartConditionMap, subChartConditionMap, nil
-}
-
-func getAppStatus(
-	appGroup *v1alpha1.ApplicationGroup,
-	chartConditionMap map[string][]metav1.Condition,
-	subChartConditionMap map[string]map[string][]metav1.Condition) []v1alpha1.ApplicationStatus {
-	// Update each application status using the HelmRelease status
-
-	var v []v1alpha1.ApplicationStatus
-	for _, app := range appGroup.Status.Applications {
-		app.ChartStatus.Conditions = chartConditionMap[app.Name]
-		for subchartName, subchartStatus := range app.Subcharts {
-			subchartStatus.Conditions = subChartConditionMap[app.Name][subchartName]
-			app.Subcharts[subchartName] = subchartStatus
-		}
-		v = append(v, app)
-	}
-	return v
-}
-
-func (r *ApplicationGroupReconciler) rollbackFailedHelmReleases(ctx context.Context, hrs []fluxhelmv2beta1.HelmRelease) error {
-	for _, hr := range hrs {
-		err := utils.HelmRollback(hr.Spec.ReleaseName, hr.Spec.TargetNamespace)
-		if err != nil {
-			return err
-		}
-		err = r.Client.Delete(ctx, &hr)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr logr.Logger, g v1alpha1.ApplicationGroup) bool {
-	nodes := make(map[string]v1alpha12.NodeStatus)
-	wfs := v1alpha12.WorkflowList{}
-	listOption := client.MatchingLabels{
-		workflow.OwnershipLabel: g.Name,
-		workflow.HeritageLabel:  workflow.Project,
-	}
-	_ = r.List(ctx, &wfs, listOption)
-
-	if wfs.Items.Len() != 0 {
-		wf := wfs.Items[0]
-		// suspend the forward workflow if it is still running
-		err := r.suspendWorkflow(ctx, logr, &wf)
-		if err != nil {
-			logr.Error(err, "failed to suspend forward workflow")
-			return false
-		}
-		for _, node := range wf.Status.Nodes {
-			nodes[node.ID] = node
+	var requeueDuration time.Duration
+	var shouldRemediate bool
+	var err error
+	// While ready is progressing, we get the state of the workflow
+	if appGroup.Generation != appGroup.Status.LastSucceededGeneration {
+		if err := statusHelper.UpdateStatus(ctx, appGroup); err != nil {
```

Comment:
This is common to both the if and else blocks so we can do this before the conditional.